pyccel/codegen/printing/fcode.py

```diff
@@ -791,32 +831,56 @@ def _get_print_format_and_arg(self,var):
         -------
         arg_format : str
-                     The format string
-        arg        : str
-                     The fortran code which represents var
+            The format string.
+        arg : str
+            The fortran code which represents var.
         """
+        if var_code is None:
+            var_code = self._print(var)
+        arg = var_code
+
         var_type = var.dtype
-        try:
-            arg_format = type_to_print_format[str(var_type)]
-        except KeyError:
-            errors.report("{} type is not supported currently".format(var_type), severity='fatal')
-
-        if var_type is NativeComplex():
-            arg = '{}, {}'.format(self._print(NumpyReal(var)), self._print(NumpyImag(var)))
-        elif var_type is NativeBool():
-            if isinstance(var, LiteralTrue):
-                arg = "'True'"
-            elif isinstance(var, LiteralFalse):
-                arg = "'False'"
+        if isinstance(var.class_type, StringType):
+            arg_format = 'A'
+        elif isinstance(var, FunctionCall) and len(var.funcdef.results)>1 or \
+                isinstance(var.class_type, InhomogeneousTupleType):
+            var_elem_code = var_code[1:-1].split(', ')
+            args_and_formats = [self._get_print_format_and_arg(v.var, c) for v,c in zip(var.funcdef.results, var_elem_code)]
+            formats = ',", ",'.join(af[0] for af in args_and_formats)
+            arg_format = f'"(",{formats},")"'
+        elif isinstance(var_type, FixedSizeNumericType):
+            if isinstance(var_type.primitive_type, PrimitiveComplexType):
+                float_format, real_arg = self._get_print_format_and_arg(NumpyReal(var))
+                imag_arg = self._print(NumpyImag(var))
+                arg_format = f'"(",{float_format}," + ",{float_format},"j)"'
+                arg = f'{real_arg}, {imag_arg}'
+            elif isinstance(var_type.primitive_type, PrimitiveFloatingPointType):
+                resolution = np.finfo(pyccel_type_to_original_type[var_type]).resolution
+                dps = int(-np.log10(resolution))
+                arg_format = f'F0.{dps}'
+            elif isinstance(var_type.primitive_type, PrimitiveIntegerType):
+                arg_format = 'I0'
+            elif isinstance(var_type.primitive_type, PrimitiveBooleanType):
+                arg_format = 'A'
+                if isinstance(var, LiteralTrue):
+                    arg = "'True'"
+                elif isinstance(var, LiteralFalse):
+                    arg = "'False'"
+                else:
+                    arg = f'merge("True ", "False", {var_code})'
             else:
-                arg = 'merge("True ", "False", {})'.format(self._print(var))
+                errors.report(f"Printing {var_type} type is not supported currently", severity='fatal')
         else:
-            arg = self._print(var)
+            errors.report(f"Printing {var_type} type is not supported currently", severity='fatal')
 
```

Comment:
This code is much longer than the original one! Is this a byproduct of the new datatyping system, or have you added more cases which were not originally covered? In the latter case please mention this in the PR description.

Reply:
Huh, I forgot that I changed this. The change isn't quite as long as it looks. There is also this code that is removed:
https://github.com/pyccel/pyccel/blob/6c5b5373b711e9ce2066a48e773625c05054a390/pyccel/codegen/printing/fcode.py#L177-L184

But the code is improved. Previously we were printing floats and complexes with 12 decimals of precision. The updated version of the code uses the precision of the float/complex to only print the decimals that are actually stored. (Minor simplification 935b37931).
Once I changed the handling for floats I realised that strings and integers were the only objects in the dictionary which weren't handled as a special case so I didn't think the dictionary was particularly useful.

The case of a function returning 2 values with different types is also handled.

I have updated the CHANGELOG : cd7c39b6a
And added a test for functions : 59fc8cd10

Reply:
PR description updated